### PR TITLE
API Generated files API

### PR DESCRIPTION
**Pull request by tractorcow**
This feature is rebased against the open asset refactor pull requests:
 - https://github.com/silverstripe/silverstripe-framework/pull/4679
 - https://github.com/silverstripe/silverstripe-cms/pull/1291
 
 This feature replaces the old requirements combination API and introduces a new generated files handler. The Requirements API has been updated quite a bit (better coding conventions, error checking). JS minification has been removed from combined files, since there are already thousands of existing front-end minification tools available that don't rely on server side minification.
 
 This is also used by the ErrorPage handler, which is refactored into a separate error handler in an upcoming pull request.

### CODE REVIEW

**Comment by tractorcow** [comment]
Rebased since the parent bugfix was merged. :)

**Comment by chillu** [comment]
> JS minification has been removed from combined files, since there are already thousands of existing front-end minification tools available that don't rely on server side minification.

This will increase the size of CMS combined files quite a bit. Even if you account for GZIP transfer, it's 367KB rather than 195KB for the three CMS combined files. Without GZIP, its 1500KB rather than 850KB. Better frontend tooling is currently in the "maybe" pile for 4.0, so for now I'd consider removing JSMin a performance regression.

File: docs/en/04_Changelogs/4.0.0.md [hunk 1]
```diff
@@ -26,6 +28,8 @@
  * `ShortcodeHandler` interface to help generate standard handlers for HTML shortcodes in the editor.
  * `AssetNameGenerator` interface, including a `DefaultAssetNameGenerator` implementation, which is used to generate
    renaming suggestions based on an original given filename in order to resolve file duplication issues.
+ * `GeneratedAssetHandler` API now used to store and manage generated files (such as those used for error page
```
Comment (chillu) [inline docs/en/04_Changelogs/4.0.0.md]: Needs to mention all the changed method names for `Requirements`, e.g. `clear_combined_files()` -> `clearCombinedFiles()`

File: view/Requirements.php [hunk 2]
```diff
@@ -1034,269 +1165,275 @@ protected function path_for_file($fileOrUrl) {
...
+	}
 
-				$combinedData .= $fileContent . "\n";
-			}
+	/**
+	 * Check if combined files are enabled
+	 *
+	 * @return bool
+	 */
+	protected function enabledCombinedFiles() {
```
Comment (chillu) [inline view/Requirements.php]: Why is this worded the other way around? `enabledCombinedFiles` vs `combinedFilesEnabled`

File: filesystem/storage/CacheGeneratedAssetHandler.php [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+		$cache = SS_Cache::factory('CacheGeneratedAssetHandler');
+		$cache->setLifetime(Config::inst()->get(__CLASS__, 'lifetime'));
+		return $cache;
+	}
+
+	/**
+	 * Flush the cache
+	 */
+	public static function flush() {
+		self::get_cache()->clean();
```
Comment (chillu) [inline filesystem/storage/CacheGeneratedAssetHandler.php]: I'd expect this to delete the underlying files as well, not only the cached references to them

Comment (tractorcow) [inline filesystem/storage/CacheGeneratedAssetHandler.php]: Ah, well, there is no "delete" method in the APL yet. I'll add a "todo" :D

**Comment by chillu** [comment]
How are we handling file deletion here? At the moment, any change to underlying files will create a new file record. Depending on how often you deploy, those can pile up. Example structure after two edits to `LeftAndMain.js`

```
./_combinedfiles/07a900d56b/leftandmain.js
./_combinedfiles/43d7f16f64/cmsmain.js
./_combinedfiles/777d368a44/leftandmain.js
./_combinedfiles/bbb1272b58/lib.js
./_combinedfiles/fcf76a715a/leftandmain.js
``

Do you see this being solved by the [file deletion discussion](https://groups.google.com/forum/#!topic/silverstripe-dev/PihYhP3ZTN8) on the mailinglist as well? Still trying to get my head around the different options there. We could treat the `SS_Cache` entries as "references" to the file?

**Comment by tractorcow** [comment]
As soon as we can agree on a deleted files API then yes we can solve this problem. :)

File: view/Requirements.php (hunk 2 above)
Comment (tractorcow) [inline view/Requirements.php]: Ah, my phpdoc is wrong, that's why. :) There's already `getCombinedFilesEnabled`

Comment (tractorcow) [inline view/Requirements.php]: I had two methods that did the same thing. The other was getCombinedFilesEnabled, but I'll merge them. :D

**Comment by tractorcow** [comment]
@chillu I have pushed up an update:

- Minification is back. I've pushed this into an injectable `Requirements_Minifier`, so `JSMin` isn't hard coded anymore.
- Still no delete, but we'll look at this again in the future.
- Lots and lots of additional upgrading notes in the changelog, especially around the changes to ErrorPage in the CMS pull request.

File: .travis.yml [hunk 4]
```diff
@@ -17,9 +17,9 @@ env:
 
 matrix:
   include:
-    - php: 5.4
+    - php: 5.5
```
Comment (chillu) [inline .travis.yml]: We've made the decision to only support 5.5+ in 4.0 this week. But you probably shouldn't sneak this in this unrelated commit, makes it hard to trace back afterwards.

File: docs/en/04_Changelogs/4.0.0.md [hunk 5]
```diff
@@ -26,9 +26,22 @@
...
+   cache or combined files).
+ * `Requirements_Midifier` API can be used to declare any new mechanism for minifying combined required files.
```
Comment (chillu) [inline docs/en/04_Changelogs/4.0.0.md]: Misspelled `Minifier`

File: docs/en/04_Changelogs/4.0.0.md [hunk 6]
```diff
@@ -74,6 +87,44 @@ The following filesystem synchronisation methods are also removed
 
 The Spyc YAML library has been removed from /thirdparty. Please load it yourself, or use the Symfony YAML component that’s automatically installed by composer.
 
+### Requirements
+
+The following methods and properties on `Requirements_Backend` have been renamed:
+
+ * `Requirements_Backund::$combine_files` made protected and renamed `$combinedFiles`
+ * `Requirements_Backend::$combine_js_with_min` made protected and renamed `$minifyCombinedFiles`
```
Comment (chillu) [inline docs/en/04_Changelogs/4.0.0.md]: Spelling - should read `combine_js_with_jsmin`

**Comment by chillu** [comment]
I've pushed these changes now - waiting for build to pass before merging

**Comment by chillu** [comment]
Heh OK, so the PR implicitly relies on the PHP 5.5 increment. I still think it's not the right place to commit this alongside other work, so created separate PRs for it: https://github.com/silverstripe/silverstripe-framework/pull/4694 and https://github.com/silverstripe/silverstripe-cms/pull/1303
